### PR TITLE
[BEAM-10616] Added Python Pardo load tests for streaming on Dataflow

### PR DESCRIPTION
**Pull request by kkucharc**
Added streaming tests cases for load tests for Python SDK. It required to parametrise tests and add condition to set specific source for Dataflow tests (Flink uses `StatefulLoadGenerator` and Dataflow `SyntheticSources`).

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by kkucharc** [comment]
run seed job

**Comment by kkucharc** [comment]
run seed job

**Comment by kkucharc** [comment]
Run Python Load Tests ParDo Flink Streaming

**Comment by kkucharc** [comment]
Run Python Load Tests ParDo Dataflow Streaming

**Comment by kkucharc** [comment]
run seed job

**Comment by kkucharc** [comment]
run seed job

**Comment by kkucharc** [comment]
Run Python Load Tests ParDo Dataflow Streaming

**Comment by kkucharc** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by kkucharc** [comment]
Run Python Load Tests ParDo Dataflow Batch

**Comment by kkucharc** [comment]
Run Python Load Tests ParDo Flink Streaming

**Comment by kkucharc** [comment]
run seed job

**Comment by kkucharc** [comment]
Run Python Load Tests ParDo Dataflow Streaming

**Comment by kkucharc** [comment]
run seed job

**Comment by kkucharc** [comment]
Run Python Load Tests ParDo Dataflow Streaming

**Comment by kkucharc** [comment]
Run Python Load Tests ParDo Dataflow Batch

**Comment by kkucharc** [comment]
Run Python Load Tests ParDo Flink Streaming

**Comment by kkucharc** [comment]
r: @tysonjh would you find time to take a look?

File: .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy [hunk 1]
```diff
@@ -151,3 +151,35 @@ CronJobBuilder.cronJob('beam_LoadTests_Python_ParDo_Dataflow_Batch', 'H 13 * * *
...
   batchLoadTestJob(delegate, CommonTestProperties.TriggeringContext.POST_COMMIT)
 }
+
+def streamingLoadTestJob = { scope, triggeringContext ->
+  scope.description('Runs Python ParDo load tests on Dataflow runner in streaming mode')
+  commonJobProperties.setTopLevelMainJobProperties(scope, 'master', 120)
+
+  def datasetName = loadTestsBuilder.getBigQueryDataset('load_test', triggeringContext)
+  for (testConfiguration in loadTestConfigurations("streaming", datasetName)) {
+    // Skipping case 2 in streaming because it timeouts. To be checked TODO: kkucharc
```
[the remark this answers is not in the document]
Comment (kkucharc) [reply .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy:161]: It makes sense. I saw several times TODOs with username and I thought this is convention where developers takes responsibility for this TODO. I changed it to issue.

File: sdks/python/apache_beam/testing/load_tests/pardo_test.py [hunk 2]
```diff
@@ -125,7 +125,9 @@ def process(self, element, state=state_param):
             state.add(1)
         yield element
 
-    if self.get_option_or_default('streaming', False):
+    if self.get_option_or_default(
+        'streaming',
+        False) and self.pipeline.get_option('runner') == "PortableRunner":
```
[the remark this answers is not in the document]
Comment (kkucharc) [reply sdks/python/apache_beam/testing/load_tests/pardo_test.py:130]: The PortableRunner is supposed to use `StatefulLoadGenerator` to enable streaming.
To be honest I assumed that if we are comparing core operations, we should be using the same source method -  SyntheticSources. 
On the other hand maybe `StatefulLoadGenerator` should be part of SyntheticSources. WDYT @tysonjh ?

[the remark this answers is not in the document]
Comment (tysonjh) [reply sdks/python/apache_beam/testing/load_tests/pardo_test.py:130]: Right, I remember you mentioned there was some issue with the SyntheticSource that Max ran into. Ideally Synthetic source would work for all standard uses but I'm not familiar with the details. How big of a change would it be to fix SyntheticSource?

I'm OK with moving forward with this change, creating a Jira issue for tracking the SyntheticSource improvement, and noting it here as a TODO.

File: .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy [hunk 3]
```diff
@@ -151,3 +151,35 @@ CronJobBuilder.cronJob('beam_LoadTests_Python_ParDo_Dataflow_Batch', 'H 13 * * *
...
+    if(testConfiguration.title != "ParDo Python Load test: 2GB 100 byte records 200 times") {
```
[the remark this answers is not in the document]
Comment (kkucharc) [reply .test-infra/jenkins/job_LoadTests_ParDo_Python.groovy:162]: I am totally for refactoring those LoadTestJob methods into one. Thank you for suggestion.

As it comes to excluding the test, I am usually more for implicit pointing to the test that should be excluded than by pointing ex. index of array (maybe because that index can change in case of adding new test cases).

File: sdks/python/apache_beam/testing/load_tests/pardo_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kamilwu) [reply sdks/python/apache_beam/testing/load_tests/pardo_test.py:147]: Actually, `StatefulLoadGenerator` is no longer required to enable streaming for Flink (please do note I'm using `Flink` instead of `PortableRunner`. `PortableRunner` can run pipelines on any engine that supports portability, Flink included). SyntheticSource executes as SDF (Splittable DoFn). Since Python SDK supports streaming SDF (https://issues.apache.org/jira/browse/BEAM-3742), SyntheticSource should work on Flink.

@mxm We'd love to hear your opinion. Should streaming ParDo tests for Flink still use `StatefulLoadGenerator`?

[the remark this answers is not in the document]
Comment (mxm) [reply sdks/python/apache_beam/testing/load_tests/pardo_test.py:147]: It's definitely not ideal to have two different sources here. Ideally, we would just use `SyntheticSource` for all tests. If we can replicate the behavior of `StatefulLoadGenerator` with `SyntheticSource`, it would be great to remove the former.

[the remark this answers is not in the document]
Comment (kamilwu) [reply sdks/python/apache_beam/testing/load_tests/pardo_test.py:147]: I did some more tests in streaming on Flink and it turned out that `SyntheticSource` is slower than `StatefulLoadGenerator`. I think it's because `SyntheticSource` is not able to use the benefits of multiple workers (only one task manager generates data).

That being said, I think we should keep two different sources for now. As for the `if` statement above, in my opinion, a new switch (for example, `--use_stateful_load_generator)` be a good idea.

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Streaming

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Batch

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Flink Streaming

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Flink Streaming

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Batch

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Streaming

**Comment by kamilwu** [comment]
@tysonjh Could you take a look once again? 

By the way, Kasia in on vacation and I'm taking care of her PR now.

**Comment by kkucharc** [comment]
@tysonjh I applied all of suggestions. I also checked the timeouting test case - adding any condition or even new variable with None value to tested ParDo is extending runtime significantly. Right now Flink and Dataflow have dedicated ParDo operation and all of tests are passing.

**Comment by kkucharc** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
R: @tysonjh

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Streaming

**Review by tysonjh — APPROVED** [review]
LGTM thanks!

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Python Load Tests ParDo Dataflow Streaming

**Comment by kamilwu** [comment]
Run PythonFormatter PreCommit

**Comment by kkucharc** [comment]
Thank you for review @tysonjh and contribution @kamilwu ! I'm merging it then.
